container/src/container_state.rs

```diff
@@ -0,0 +1,187 @@
+use crate::{Error, Result};
+use serde::{Deserialize, Serialize};
+use std::fs;
+use std::fs::{File, OpenOptions};
+use std::io::Write;
+use std::path::PathBuf;
+
+const KAPS_ROOT_PATH: &str = "/var/run/kaps";
+const OCI_VERSION: &str = "0.2.0";
+const STATE_FILE: &str = "state.json";
+
+/// Container runtime status
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+pub enum Status {
+    #[serde(rename = "creating")]
+    Creating,
+    #[serde(rename = "created")]
+    Created,
+    #[serde(rename = "running")]
+    Running,
+    #[serde(rename = "stopped")]
+    Stopped,
+}
+
+impl Default for Status {
+    fn default() -> Self {
+        Status::Creating
+    }
+}
+
+/// Represent the state of the running container.
+#[derive(Serialize, Deserialize)]
+pub struct ContainerState {
+    id: String,
+    /// Path where the state of the container will be saved.
+    #[serde(skip)]
+    container_path: PathBuf,
+    /// OCI version.
+    oci_version: String,
+    /// Runtime state of the container.
+    pub status: Status,
+    /// ID of the container process.
+    pub pid: i32,
+    /// Path to the bundle.
+    bundle: PathBuf,
+}
+
+impl Default for ContainerState {
+    fn default() -> Self {
+        ContainerState {
+            container_path: PathBuf::default(),
+            oci_version: OCI_VERSION.to_string(),
+            id: String::default(),
+            status: Status::default(),
+            pid: 0,
+            bundle: PathBuf::default(),
+        }
+    }
+}
+
+impl ContainerState {
+    pub fn new(id: &str, bundle_path: &str, container_dir: Option<&str>) -> Result<Self> {
```

Comment:
There should not be a `container_dir` option. All container statuses should be stored under a fixed path, `/var/run/kaps0/containers/` for example.

Reply:
I was using this parameter to change the path for container status in the test module.

Do the tests have to use the same path or it is a good thing to have a test path?  

If this is the case have you an alternative to change the state path dynamically between test and reel usage that is different from using the `container_dir` parameter?

Reply:
> I was using this parameter to change the path for container status in the test module.
> 
> Do the tests have to use the same path or it is a good thing to have a test path?

It's much better to have a test path.

> If this is the case have you an alternative to change the state path dynamically between test and reel usage that is different from using the `container_dir` parameter?

What really matters is your public API, so you could go with the following pattern:

```rust
    pub fn new(id: &str, bundle_path: &str) -> Result<Self> {
        ContainerState::_new(id, bundle_path, KAPS_ROOT_PATH)
    }

    fn _new(id: &str, bundle_path: &str, container_dir: &str) -> Result<Self> {
        let bundle: PathBuf = PathBuf::from(bundle_path);
        let container_path: PathBuf = PathBuf::from(container_dir).join(id);

        if container_path.as_path().exists() {
            return Err(Error::ContainerExists(format!(
                "A container with the id '{}' already exists",
                id
            )));
        }

        // create the container directory
        fs::create_dir_all(&container_path).map_err(|e: Error| Error::CreateStateFile(e.to_string()))?;

        // create the `state.json` file of the container
        File::create(container_path.join(STATE_FILE))
            .map_err(|e: Error| Error::CreateStateFile(e.to_string()))?;

        Ok(ContainerState {
            container_path,
            id: id.to_string(),
            bundle,
            ..Default::default()
        })
    }
```

And use either the pivate `_new` or the public `new` API from your tests.